### PR TITLE
Add env var to set DEFAULT_PYTHON (add Pyenv support)

### LINKED ISSUE #113: Add pyenv compatibility
I've got a CLI-tool (tmux-res) I built for handling some common tasks in my TMUX-environment which I built a while back. For fun I wanted to try installing it via pipx but it doesn't work.

This is the package:
https://pypi.org/project/tmux-res/

```sh
 $ pyenv global 3.7.2
 $ python --version
Python 3.7.2

 $ pipx install tmux-res
  Could not find a version that satisfies the requirement tmux-res (from versions: )
No matching distribution found for tmux-res
'/Users/hultner/.local/pipx/venvs/tmux-res/bin/python -m pip install tmux-res -q' failed

 $ pip install tmux-res
Collecting tmux-res
  Downloading https://files.pythonhosted.org/packages/09/80/a0294c244aebbdc9f13c06b182a7c1857618da20eabd1c88c4f449dbdf09/tmux_res-0.4.8-py3-none-any.whl
Collecting python-dotenv (from tmux-res)
  Downloading https://files.pythonhosted.org/packages/8c/14/501508b016e7b1ad0eb91bba581e66ad9bfc7c66fcacbb580eaf9bc38458/python_dotenv-0.10.1-py2.py3-none-any.whl
Collecting Click (from tmux-res)
  Using cached https://files.pythonhosted.org/packages/fa/37/45185cb5abbc30d7257104c434fe0b07e5a195a6847506c074527aa599ec/Click-7.0-py2.py3-none-any.whl
Installing collected packages: python-dotenv, Click, tmux-res
Successfully installed Click-7.0 python-dotenv-0.10.1 tmux-res-0.4.8
You are using pip version 18.1, however version 19.0.3 is available.
You should consider upgrading via the 'pip install --upgrade pip' command.
```

### PR DESCRIPTION
**Pull request by wren**
Fixes #113.

This keeps sys.executable as the default, but will use the PIPX_DEFAULT_PYTHON environment variable, if defined. This mean no change in functionality for any users until they define the variable.

I'm suggesting the var `PIPX_DEFAULT_PYTHON` (no `PYENV` even though the original issue asked for pyenv support), because this approach will work with any python manager, and I think it might be confusing to enshrine one manager in the var name over others.

Also, it didn't look like the constants are tested, so I didn't update any tests. But let me know if I missed something and I'll be happy to add.

### CODE REVIEW

**Comment by cs01** [comment]
I am not a pyenv user (maybe one of the other pipx maintainers is?), so I don't really know how to test this or show that it fixes anything. 

Could you run some commands that show this fixes something, and copy+paste them here so I can see what that looks like?

Assuming this change is useful and we want to merge, the help text will also need to document this environment variable (see `main.py`).

**Comment by wren** [comment]
Generally (for any user): the new env var should let you override the default python used by pipx to install packages. So, if you `export PIPX_DEFAULT_PYTHON=python3.6`, then pipx will use `python3.6` to install by default (but is still override-able by `--python`). 

Specifically (for Pyenv users): this works for with `export PIPX_DEFAULT_PYTHON="$PYENV_ROOT/shims/python"` since Pyenv uses that shim to always point to the correct Python (note: I think `PYENV_ROOT` defaults to `~/.pyenv`).

So, this is testable without Pyenv simply by changing the env var and verifying that Pipx is using the correct Python to install packages.

## Pyenv Tests
I tested this by running the latest pipx release side-by-side with this the source from this PR.

https://asciinema.org/a/347120

```sh
$ pyenv global 3.8.3

$ python --version
Python 3.8.3

$ env | grep -i pipx_default_python
PIPX_DEFAULT_PYTHON=/Users/jonathan/.cache/pyenv/shims/python

$ python src/pipx install jrnl
  installed package jrnl 2.4.3, Python 3.8.3
  These apps are now globally available
    - jrnl
done! ✨ 🌟 ✨

$ python src/pipx uninstall jrnl
uninstalled jrnl! ✨ 🌟 ✨

$ pipx install jrnl
  installed package jrnl 2.4.3, Python 3.8.3
  These apps are now globally available
    - jrnl
done! ✨ 🌟 ✨

$ pipx uninstall jrnl
uninstalled jrnl! ✨ 🌟 ✨

$ pyenv global 3.6.10

$ python --version
Python 3.6.10

$ python src/pipx install jrnl
  installed package jrnl 2.4.3, Python 3.6.10
  These apps are now globally available
    - jrnl
done! ✨ 🌟 ✨

$ python src/pipx uninstall jrnl
uninstalled jrnl! ✨ 🌟 ✨

$ pipx install jrnl
  installed package jrnl 2.4.3, Python 3.8.3
  These apps are now globally available
    - jrnl
done! ✨ 🌟 ✨

$ pipx uninstall jrnl
uninstalled jrnl! ✨ 🌟 ✨

```

**Comment by uranusjr** [comment]
I needed to do something similar in https://github.com/pipxproject/pipx/issues/244#issuecomment-580382747. This would also be a good workaround for #395 as well. In general, `sys.executable` is not the best value in many repackaging situations, and this would be a welcoming change. And if it happens to also help pyenv users, more wonderful 😄

**Comment by cs01** [comment]
Looks good to me. I think we should add a unit test before landing. Since the tests on CI generally only have one python version installed, we can set `PIPX_DEFAULT_PYTHON` to be a path that doesn't exist, then test that pipx raises an exception saying the Python interpreter doesn't exist.

**Comment by wren** [comment]
I'm not super familiar with this project, so I'm not sure if I did what you asked, but I think this is ready now.

**Comment by uranusjr** [comment]
It just occured to me, would it be worthwhile to incorporate `shutil.which()` into the resolution so a user can specify (say) `python3` instead of `/usr/bin/python3`? It’s probably a good idea to add some error reporting around this code as well, since the exception may be quite cryptic if the user sets `DEFAULT_PYTHON` to an invalid value.

File: tests/test_constants.py
Comment (cs01) [inline tests/test_constants.py:48]: Should this be `PIPX_DEFAULT_PYTHON`?

**Comment by cs01** [comment]
> It just occured to me, would it be worthwhile to incorporate `shutil.which()` into the resolution so a user can specify (say) `python3` instead of `/usr/bin/python3`? It’s probably a good idea to add some error reporting around this code as well, since the exception may be quite cryptic if the user sets `DEFAULT_PYTHON` to an invalid value.

Agreed. We should raise a `PipxError()` with a clear error message and suggested action for the user if the python interpreter can't be found.

File: tests/test_constants.py
Comment (wren) [reply tests/test_constants.py:48]: I'm not sure, but I don't think so. The way constants are set up, environment variables are read before the tests are run, so patching the environment variable itself (`PIPX_DEFAULT_PYTHON`) wouldn't do anything here. So, we need to patch the python variable that holds the environment variable instead.

And I'm fairly new to your code base, so I might be wrong, but it seems to me this is the same way the other constants are handled in the test suite (like patching `LOCAL_BIN_DIR` instead of `PIPX_BIN_DIR` [here](https://github.com/pipxproject/pipx/blob/master/tests/conftest.py#L21)).

**Comment by wren** [comment]
Sounds good. I can add some better error output.

For the use of `shutil.which()`: I'm happy to add if that's what y'all want, but I personally don't think it's desirable behavior. It can lead to some unpredictable results as people move in and out of virtual envs, subshells, etc. I think a full path gives users more control. And if a user really wants this functionality, they can add already do something like `export PIPX_DEFAULT_PYTHON="$(which python)"` to get the same result, so I'm not sure it would add anything other than a larger maintenance burden.

**Comment by itsayellow** [comment]
> For the use of `shutil.which()`: I'm happy to add if that's what y'all want, but I personally don't think it's desirable behavior. It can lead to some unpredictable results as people move in and out of virtual envs, subshells, etc. I think a full path gives users more control. And if a user really wants this functionality, they can add already do something like `export PIPX_DEFAULT_PYTHON="$(which python)"` to get the same result, so I'm not sure it would add anything other than a larger maintenance burden.

I agree with this.  I feel like being explicit in environment config is good.  For our commands in an interactive shell it is nice that `--python` doesn't need a full path, but I don't feel the same way about env variables, and there is possibility of unpredictable results as @wren says.

**Comment by uranusjr** [comment]
Contrary to instincts, a `shutil.which()` may actually make things more explicit. Commands like `python3` would still work without `shutil.which()` since we eventually pass the value to `subprocess`, which does `PATH` resolution on the command anyway. But that resolution logic and behaviour can be different from what user expects, such as [bpo-38905](https://bugs.python.org/issue38905). `shutil.which()`, on the other hand, always behaves the same, and matches the expectation in the majority of cases.

**Comment by itsayellow** [comment]
If we do use `shutil.which` I would also suggest a `logging.info()` statement showing the full resolved path that `shutil.which` returns, so in pipx verbose mode the user (and/or pipx debugging team) knows which python is getting used.

**Comment by wren** [comment]
Hi, just checking back on this.

I just need a final decision on whether or not to include `shutil`, and I can wrap this PR up.

**Comment by itsayellow** [comment]
I'm ok with it, and would like the logging.info() statement to show that we found `PIPX_DEFAULT_PYTHON`, are using it, and what the full resolved path is.

File: tests/test_constants.py
Comment (itsayellow) [reply tests/test_constants.py:48]: I think we could do it with `importlib.reload(pipx.constants)` but I'm not sure if there would be side-effects.

**Comment by uranusjr** [comment]
The `shutil` discussion and improvement can be made separatedly. But we’ll probably need to drop the `resolve()` part in that case, since that would covert `python3` to `$PWD/python3` which would be very confusing to the user.

**Comment by cs01** [comment]
@itsayellow @uranusjr do you guys want to request changes or are you good merging like this?

**Comment by uranusjr** [comment]
I suggest improving treatment to relative paths, either by using `resolve()` to turn them absolute, or rejecting outright based on `is_absolute()`.

**Comment by itsayellow** [comment]
I suggest adding a `logging.info()` statement inside of `main.py` `setup()` after statement `logging.info(f"pipx version is {__version__}")`.
Something like:
```
logging.info(f"Default python interpreter: {constants.DEFAULT_PYTHON}")
```

**Comment by itsayellow** [comment]
In general that will make debugging much easier in the future.

**Comment by wren** [comment]
Sounds good. I'll do the requested updates this week. Thanks for the feedback!

**Comment by gaborbernat** [comment]
@wren any progress on this, you still plan to do it?

**Comment by wren** [comment]
Ah, yes. Time got away from me in the last week, sorry! I'll make some time today for it.

**Comment by itsayellow** [comment]
FYI `black` just made a new release hours ago, which is likely why your Lint test is failing.

I'm going to see if I can make a merge that will fix things soon.

**Comment by itsayellow** [comment]
OK, if you merge pipx master now the tests *should* work again.

**Comment by wren** [comment]
I just came back to check on this and realized I didn't hit submit on my comment. Sorry!

--

So, I've made the requested changes. I also added a call to `shutil.which` because it was the simplest way to deal with the relative path handling requirement.

The only test failing is related to typing (since the variable can now be `None` in addition to a string). I can fix this either by updating the type hints in the relevant functions, or by forcing a string. Either way is fine with me. Do y'all have a preference?

**Comment by uranusjr** [comment]
I think it’s reasonable to default to sys.executable when which returns None, i.e. `which(...) or sys.executable`.

**Comment by wren** [comment]
@uranusjr I'm okay with that approach. Just to be clear, though, that would make it impossible to give an error if the env var is invalid since it's immediately with the sys python. Is that okay?

I could also move the `which` call into main.py, and then overwrite `DEFAULT_PYTHON` with the absolute path after it's been validated by `which` (and the relevant error has been thrown). Would you prefer that?

**Comment by uranusjr** [comment]
Ah, I did not think about the error reporting issue if the environment variable is invalid, thanks for the reminder. Maybe we should generate a warning in that case instead. Following the approach of #465 (which moves `DEFAULT_PYTHON` to its own module and would conflict with this PR), we should probably refactor this logic into a function, and raise a `PipxError` is the environment variable is invalid instead.

**Comment by wren** [comment]
Hi, this is updated now. I believe this meets all of the previously established criteria:

- `DEFAULT_PYTHON` is always a string (so no updating of type hints is necessary)
- Supports the new Windows embeddable python code (#465)
- Any env var setting is validated with `shutil.which`
- Sys executable (including the new windows one) is not run through which (since it's redundant)
- A `PipxError` is raised when `which` fails
- `DEFAULT_PYTHON` is updated with an absolute path if validation succeeds
- logging statement in place for help debugging
- The env var can still be overridden by passing `--python` on cli

Anything I've left out?

File: src/pipx/interpreter.py
Comment (uranusjr) [inline src/pipx/interpreter.py:70]: `{env_python!r}` or `{repr(env_python)}` would handle the quoting edge cases better.

**Review by uranusjr — COMMENTED** [review]
One minor nitpick, otherwise this looks good to me!

File: src/pipx/interpreter.py
Comment (wren) [reply src/pipx/interpreter.py:70]: Hi, would you mind giving an example of a quoting edge case that `repr` would fix? I thought that f-strings already handled escaping, and since `env_python` is always a string, `repr` seems redundant to me. Happy to add either way, but I just want to understand if I'm missing something.

Comment (uranusjr) [reply src/pipx/interpreter.py:70]: The main difference is when the path contains a quote character. `repr()` would automatically choose the best quoting scheme, and perform better than the current naive method of simply adding quote characters around the path.

Alternatively you can just not quote the path, like you did for the “python interpreter is” message below. That would cause some confusion when there’s space in the path, but there’s no perfect solution here, unfortunately.

Comment (wren) [reply src/pipx/interpreter.py:70]: Ah, okay. Thank you for explaining!

**Review by uranusjr — APPROVED** [review]

File: src/pipx/interpreter.py
Comment (itsayellow) [inline src/pipx/interpreter.py:74]: Since `env_default_python` is not used anywhere else but this `if` statement, could you combine lines 74,76 into one line for code simplicity and readability:
```
if not os.environ.get("PIPX_DEFAULT_PYTHON"):
```

**Review by itsayellow — CHANGES_REQUESTED** [review]
Looks mostly good to me, just one little nit.

File: src/pipx/interpreter.py
Comment (wren) [reply src/pipx/interpreter.py:74]: It's used in two places below, which is why I assigned it to a variable instead of calling `os.environ` twice. Do you still want me to get rid of it?

Comment (itsayellow) [reply src/pipx/interpreter.py:74]: Oh sorry I see inside the else.  Ok you can forget this.

**Comment by itsayellow** [comment]
Hey @wren , just in case it got lost, could you please list PIPX_DEFAULT_PYTHON and a help blurb for it in main.py around line 50, inside PIPX_DESCRIPTION, under "Optional Environment Variables"?  Then I'll merge this.

**Comment by wren** [comment]
@itsayellow Yup, sorry. I'm in the middle of moving right now, so it'll be a few days before I can get to this.

**Comment by wren** [comment]
@itsayellow All ready for merging. Thanks for your patience!

**Comment by itsayellow** [comment]
@wren, thanks for your persistence as well!

**Review by itsayellow — APPROVED** [review]

**Review by gaborbernat — APPROVED** [review]

**Comment by itsayellow** [comment]
Just one more thing and I'll do it...updating changelog entry.

**Pull request merged**
